Mathlib/Tactic/DeriveToExpr.lean

```diff
@@ -0,0 +1,216 @@
+/-
+Copyright (c) 2023 Kyle Miller. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kyle Miller
+-/
+import Lean
+
+/-!
+# A `ToExpr` derive handler
+
+This module defines a `ToExpr` derive handler for inductive types. It supports mutually inductive
+types as well.
+
+It is originally based on the `Repr` derived handler from Lean 4 itself.
+
+The `ToExpr` derive handlers support universe level polymorphism. This is implemented using the
+`Lean.ToLevel` class, where `Lean.ToLevel.toLevel (Sort u)` evaluates to a `Lean.Level` term
+representing `u`. To use `ToExpr` in places where there is universe polymorphism, make sure
+to have a `[ToLevel (Sort u)]` instance available.
+-/
+
+namespace Lean
+/-- Given a type universe `Sort u`, produce a `Level` expression that denotes the level `u`.
+
+One reason this takes an argument rather than taking just a level parameter is
+to avoid the "unused universe parameter" error. -/
```

Comment:
Could we add a `PUnit.{u}` field instead, which would then directly align with the lean3 version?

Reply:
Is that a compromise we need to make? It's kind of neat how you can write `ToLevel (Sort u)` or `ToLevel (Type u)`.  The linter is also happy with this class as-is. There seem to be only four uses of `reflected_univ` in mathlib 3 outside of `src/meta/univs.lean`.

Reply:
Doesn't `ToLevel (Sort u)` mean `ToLevel.{u+1} (Sort u)`? That seems a bit confusing to me. Adding the punit field will pretty much be un-noticeable to the caller, so I think the only concern is the +1.

Reply:
> There seem to be only four uses of `reflected_univ` in mathlib 3 outside of `src/meta/univs.lean`.

Note that the one in `Data.Matrix.Notation` is ready for porting; for now I've left a note suggesting it should wait till this PR is done.

https://github.com/leanprover-community/mathlib/blob/d95bef0d215ea58c0fd7bbc4b151bf3fe952c095/src/data/matrix/notation.lean#L50-L57

Reply:
(just to be clear, I agree that perfect alignment with Lean3 is a non-issue here)

Reply:
I wasn't thinking correctly about `ToLevel (Type u)`, which isn't very useful. I switched to `ToLevel.{u}` (using a `Type u` to use `u` rather than `PUnit`).